fms/utils/generation.py

```diff
@@ -126,3 +128,234 @@ def truncate_after_eos(result, eos_token_id):
         eos_idx = eos_idx[0].item()
         result = result[: eos_idx + 1]
     return result
+
+
+def speculative_generate(
+    model: Union[Callable, torch.nn.Module],
+    input_ids: List[torch.Tensor],
+    speculator: Speculator,
+    max_seq_len: int = 2048,
+    new_tokens: int = 256,
+    top_k: int = 5,
+    threshes=[5, 3, 2],
+    verbose_dict=None,
```

Comment:
this should probably have a more complete type annotation so it's clearer how to use it (e.g. Optional[Mapping[ ??, ??]]` ?)